src/Inciter/Diagnostics.C

```diff
@@ -0,0 +1,141 @@
+/*!
+  \file      src/Inciter/Diagnostics.C
+  \copyright 2012-2015, J. Bakosi, 2016-2017, Los Alamos National Security, LLC.
+  \brief     Diagnostics class for collecting diagnostics
+  \details   Diagnostics class for collecting diagnostics, e.g., residuals, and
+    various norms of errors while solving partial differential equations.
+*/
+// *****************************************************************************
+
+#include "PDE.h"
+#include "Diagnostics.h"
+#include "DiagReducer.h"
+#include "Discretization.h"
+#include "Inciter/InputDeck/InputDeck.h"
+
+namespace inciter {
+
+extern ctr::InputDeck g_inputdeck;
+extern std::vector< PDE > g_pdes;
+
+static CkReduction::reducerType DiagMerger;
+
+} // inciter::
+
+using inciter::Diagnostics;
+
+Diagnostics::Diagnostics( const Discretization& d )
+// *****************************************************************************
+//  Compute diagnostics, e.g., residuals, norms of errors, etc.
+//! \param[in] d Discretization proxy to read from
+// *****************************************************************************
+{
+  // Store the local IDs of those mesh nodes to which we contribute but do not
+  // own, i.e., slave nodes. Ownership here is defined by having a lower chare
+  // ID than any other chare that also contributes to the node.
+  for (const auto& c : d.Msum())        // for all chares that neighbor our mesh
+    if (d.thisIndex > c.first)          // if our chare ID is larger than theirs
+      for (auto i : c.second)           // store local ID in set
+        m_slave.insert( tk::cref_find( d.Lid(), i ) );
+}
+
+void
+Diagnostics::registerReducers()
+// *****************************************************************************
+//  Configure Charm++ reduction types
+//! \details This routine is supposed to be called from a Charm++ nodeinit
+//!   routine. Since the runtime system executes nodeinit routines exactly once
+//!   on every logical node early on in the Charm++ init sequence, they must be
+//!   static as they are called without an object. See also: Section
+//!   "Initializations at Program Startup" at in the Charm++ manual
+//!   http://charm.cs.illinois.edu/manuals/html/charm++/manual.html.
+// *****************************************************************************
+{
+  DiagMerger = CkReduction::addReducer( tk::mergeDiag );
+}
+
+bool
+Diagnostics::compute( Discretization& d, const tk::Fields& u )
+// *****************************************************************************
+//  Compute diagnostics, e.g., residuals, norms of errors, etc.
+//! \param[in] d Discretization proxy to read from
+//! \param[in] u Current solution vector
+//! \return True if diagnostics have been computed
+// *****************************************************************************
+{
+  // Optionally collect diagnostics and send for aggregation across all workers
+  if ( !((d.It()+1) % g_inputdeck.get< tag::interval, tag::diag >()) ) {
```

Comment:
Would the clarity of this be increased by breaking it out into multiple lines?

Reply:
Probably. I would do something like:
```
// Optionally collect diagnostics and send for aggregation across all workers
auto diag = g_inputdeck.get< tag::interval, tag::diag >();
if ( !((d.It()+1) % diag) ) {
```
What do you think?

Reply:
Probably. I would do something like:
```
auto diag = g_inputdeck.get< tag::interval, tag::diag >();
if ( !((d.It()+1) % diag) ) {
...
}
```
What do you think?

Reply:
I like it more.

I'd personally be more explicit about the condition you're negating. At a glance when I see this i expect something like if `!(12 % 10) => if !(2)` 

If we're trying to see if the % == 0 or similar, I'd prefer to explicitly state it?

Reply:
Okay, sure. There is another place also where I have done this. I will fix this up tomorrow. Thx.

Reply:
Fixed: 26ac6821

Reply:
Fixed: 26ac6821